odl/tomo/backends/astra_cuda.py

```diff
@@ -38,207 +38,340 @@
+        elif self.geometry.ndim == 3:
+            if out is None:
+                out = self.proj_space.element()
+
+            out[:] = np.rollaxis(self.out_array, 0, 3)
+        else:
+            raise RuntimeError('unknown ndim')
+
+        # Fix inconsistent scaling
+        if isinstance(self.geometry, Parallel2dGeometry):
+            # parallel2d scales with pixel stride
+            out *= 1 / float(self.geometry.det_partition.cell_sides[0])
+
+        if not self.use_cache:
+            self.delete_ids()
+
+        return out
+
+    def create_ids(self):
+        ndim = self.geometry.ndim
+
+        # Create input and output arrays
+        if ndim == 2:
+            astra_proj_shape = self.proj_space.shape
+        elif ndim == 3:
+            astra_proj_shape = (self.proj_space.shape[2],
+                                self.proj_space.shape[0],
+                                self.proj_space.shape[1])
+
+        self.in_array = np.empty(self.reco_space.shape,
+                                 dtype='float32', order='C')
+        self.out_array = np.empty(astra_proj_shape,
+                                  dtype='float32', order='C')
+
+        # Create ASTRA data structures
+        vol_geom = astra_volume_geometry(self.reco_space)
+        proj_geom = astra_projection_geometry(self.geometry)
+        self.vol_id = astra_data(vol_geom, datatype='volume',
+                                 data=self.in_array, allow_copy=False)
+
+        self.proj_id = astra_projector('nearest', vol_geom, proj_geom,
+                                       ndim, impl='cuda')
+
+        self.sino_id = astra_data(proj_geom,
+                                  datatype='projection',
+                                  ndim=self.proj_space.ndim,
+                                  data=self.out_array,
+                                  allow_copy=False)
 
         # Create algorithm
-        algo_id = astra_algorithm('forward', ndim, vol_id, sino_id,
-                                  proj_id=proj_id, impl='cuda')
+        self.algo_id = astra_algorithm(
+            'forward', ndim, self.vol_id, self.sino_id,
+            proj_id=self.proj_id, impl='cuda')
+
+    def delete_ids(self):
+        """Delete ASTRA objects."""
+        if self.geometry.ndim == 2:
+            adata, aproj = astra.data2d, astra.projector
+        else:
+            adata, aproj = astra.data3d, astra.projector3d
+
+        if self.algo_id is not None:
+            astra.algorithm.delete(self.algo_id)
+            self.algo_id = None
+        if self.vol_id is not None:
+            adata.delete(self.vol_id)
+            self.vol_id = None
+        if self.sino_id is not None:
+            adata.delete(self.sino_id)
+            self.sino_id = None
+        if self.proj_id is not None:
+            aproj.delete(self.proj_id)
+            self.proj_id = None
+
+        self.in_array = None
+        self.out_array = None
+
+    def __del__(self):
+        self.delete_ids()
```

Comment:
Okay, never mind. The reference is deleted automatically when removing the object.
But then, `del` is not really recommended in the sense of "guarantee the things get deleted". 

So I would prefer having a more general `clear_cache` method that also sets the array references to `None`. It's a clearer indicator to the user that this is intended to clear the cache. You never know if `__del__` is implemented or not, but you clearly see the public methods.